File: ai-worker/app/llm.py
"""LLM narrative generation: Groq first, OpenRouter as the fallback.

The numeric fields (risk_score, fatigue_trend, focus_body_part) come from
analysis.py — deterministic and grounded. The LLM is asked only for the
qualitative pieces: a short coaching tip and a recommended action from a closed
set. This keeps the model from contradicting the math.

WHY NOT ANTHROPIC. This service used to hold its own Anthropic client while the
Cloudflare Worker and the Supabase Edge Functions each ran something else — three
AI stacks in one product, three sets of credentials, three things to rotate, and
three different ways for the same kind of failure to present. It now runs the
same chain as everything else: Groq for speed, OpenRouter behind it for breadth,
and a provider whose key is unset is skipped rather than attempted. See
supabase/functions/_shared/llm.ts, which this deliberately mirrors.

Structured output went with it. Anthropic's json_schema format has no equivalent
that every model on the chain honours, so the schema is stated in the prompt and
the reply is validated by parsing — a rung that answers with prose is a failed
rung, and the next one is tried.

With no key at all, a deterministic fallback narrative is used, so the service
runs end-to-end in dev and degrades to something sensible in production rather
than failing. That was already true and is the reason this conversion is safe.
"""

from __future__ import annotations

import json
import logging
import os
from typing import Optional

from .analysis import Features
from .models import Insight, RecommendedAction

logger = logging.getLogger("ai-worker.llm")

# Endpoint, key variable and default models per provider, in preference order.
# Both speak the OpenAI chat-completions shape, which is why this is one code
# path and not two.
_PROVIDERS: tuple[dict, ...] = (
    {
        "name": "groq",
        "url": "https://api.groq.com/openai/v1/chat/completions",
        "key_var": "GROQ_API_KEY",
        "models_var": "GROQ_TEXT_MODELS",
        "models": ("openai/gpt-oss-120b", "llama-3.3-70b-versatile"),
    },
    {
        "name": "openrouter",
        "url": "https://openrouter.ai/api/v1/chat/completions",
        "key_var": "OPENROUTER_API_KEY",
        "models_var": "OPENROUTER_TEXT_MODELS",
        "models": ("deepseek/deepseek-chat", "google/gemini-2.5-flash"),
    },
)

_TIMEOUT_S = 20.0

_ACTIONS: tuple[RecommendedAction, ...] = (
    "rest",
    "static_stretching_lower_body",
    "mobility",
    "light_recovery",
    "moderate_drill",
    "heavy_sprint",
)

# Stated in the prompt now rather than passed as an API parameter — see the
# module docstring. Only the enum is constrained; the rest is clamped in code.
_OUTPUT_SCHEMA = {
    "type": "object",
    "properties": {
        "ai_summary_text": {"type": "string"},
        "recommended_action": {"type": "string", "enum": list(_ACTIONS)},
    },
    "required": ["ai_summary_text", "recommended_action"],
    "additionalProperties": False,
}

_SYSTEM = (
    "You are a sports biomechanist and recovery coach for footballers. "
    "You are given pre-computed metrics for one athlete over a recent window. "
    "Identify the weakest link and give one specific, actionable recovery tip in "
    "at most 25 words. Be concrete (name the body part and the action). Do not "
    "invent numbers beyond those provided. "
    "Return ONLY valid minified JSON matching exactly this schema, with no prose, "
    "no markdown fence and no commentary:\n" + json.dumps(_OUTPUT_SCHEMA)
)


def _chain() -> list[tuple[str, str, str, str]]:
    """(provider, url, key, model) for each rung whose key is configured."""
    out: list[tuple[str, str, str, str]] = []
    for p in _PROVIDERS:
        key = os.environ.get(p["key_var"])
        if not key:
            continue
        configured = [m.strip() for m in os.environ.get(p["models_var"], "").split(",") if m.strip()]
        for model in (configured or list(p["models"])):
            out.append((p["name"], p["url"], key, model))
    return out


def generate_insight(features: Features, is_in_season: bool, training_note: Optional[str] = None) -> Insight:
    summary, action = _narrative(features, is_in_season, training_note)
    return Insight(
        risk_score=features.risk_score,
        fatigue_trend=features.fatigue_trend,
        ai_summary_text=summary,
        recommended_action=action,
        focus_body_part=features.focus_body_part,
    )


def _narrative(features: Features, is_in_season: bool, training_note: Optional[str] = None) -> tuple[str, RecommendedAction]:
    chain = _chain()
    if not chain:
        logger.info("no GROQ_API_KEY or OPENROUTER_API_KEY — using deterministic fallback narrative")
        return _fallback(features, is_in_season)

    prompt = _build_prompt(features, is_in_season, training_note)
    for provider, url, key, model in chain:
        label = f"{provider}/{model}"
        try:
            text = _post(url, key, model, prompt)
        except Exception:  # network/auth/rate-limit — try the next rung
            logger.warning("%s failed; trying next rung", label, exc_info=True)
            continue

        parsed = _parse(text)
        if parsed is None:
            # A model that answers with prose or an apology has not done the job.
            # Falling through beats returning it: the caller cannot tell a bad
            # answer from a good one, and the deterministic fallback below is
            # genuinely better than a confident non-answer.
            logger.warning("%s returned an unusable reply; trying next rung", label)
            continue
        return parsed

    logger.warning("every model in the chain failed; using fallback narrative")
    return _fallback(features, is_in_season)


def _post(url: str, key: str, model: str, prompt: str) -> str:
    """One chat-completions call. Raises on any non-2xx or transport error."""
    # urllib rather than a new dependency: this is one POST of JSON, and adding
    # an SDK for it would mean a third HTTP client in a service that already has
    # what it needs in the standard library.
    import urllib.error
    import urllib.request

    body = json.dumps({
        "model": model,
        "max_tokens": 1024,
        "messages": [
            {"role": "system", "content": _SYSTEM},
            {"role": "user", "content": prompt},
        ],
    }).encode()
    req = urllib.request.Request(
        url,
        data=body,
        headers={
            "Authorization": f"Bearer {key}",
            "Content-Type": "application/json",
            # OpenRouter attributes usage by these; Groq ignores what it doesn't know.
            "HTTP-Referer": "https://pocketathlete.com",
            "X-Title": "PocketAthlete",
        },
    )
    with urllib.request.urlopen(req, timeout=_TIMEOUT_S) as resp:
        payload = json.loads(resp.read())
    return payload["choices"][0]["message"]["content"]


def _parse(text: str) -> Optional[tuple[str, RecommendedAction]]:
    """The summary and action, or None if the reply isn't usable."""
    if not text or not text.strip():
        return None
    # Models wrap JSON in prose and markdown fences however firmly you ask them
    # not to, so take the outermost object rather than trusting the whole string.
    start, end = text.find("{"), text.rfind("}")
    if start < 0 or end <= start:
        return None
    try:
        data = json.loads(text[start:end + 1])
        action = data["recommended_action"]
        if action not in _ACTIONS:
            raise ValueError(f"unexpected action {action!r}")
        summary = str(data["ai_summary_text"]).strip()
        if not summary:
            raise ValueError("empty summary")
        return summary, action
    except (json.JSONDecodeError, KeyError, ValueError, TypeError):
        return None


def _build_prompt(features: Features, is_in_season: bool, training_note: Optional[str] = None) -> str:
    f = features.as_dict()
    season = "in-season (taper — favour recovery and short high-intensity work)" \
        if is_in_season else "off-season (build — heavier strength and endurance allowed)"
    training_line = f"- Training load: {training_note}\n" if training_note else ""
    return (
        f"Athlete metrics (window of {f['n_days']} days), {season}:\n"
        f"- Worst current pain: {f['max_pain_today']}/10"
        + (f" in {f['focus_body_part']}" if f["focus_body_part"] else " (none)")
        + "\n"
        f"- 7-day average worst-pain: {f['pain_7d_avg']}\n"
        f"- Sleep today: {f['sleep_today']} (z-score {f['sleep_z']})\n"
        f"- HRV today: {f['hrv_today']} (z-score {f['hrv_z']})\n"
        f"- Fatigue trend: {f['fatigue_trend']}\n"
        f"- Computed injury-risk score: {f['risk_score']} (0 safe .. 1 critical)\n"
        f"- Flags: {', '.join(f['flags']) or 'none'}\n"
        f"{training_line}\n"
        "Consider training-load progression when judging readiness. "
        "Give the recovery tip and choose the single best recommended_action."
    )


def _fallback(features: Features, is_in_season: bool) -> tuple[str, RecommendedAction]:
    """Deterministic narrative when the LLM is unavailable."""
    part = features.focus_body_part
    if features.max_pain_today >= 7 and part:
        action: RecommendedAction = "static_stretching_lower_body"
        return (
            f"{part.capitalize()} pain is high ({features.max_pain_today:.0f}/10). "
            "Skip sprints — focus on gentle mobility and static stretching there.",
            action,
        )
    if "sleep_drop" in features.flags or "hrv_drop" in features.flags:
        return (
            "Recovery markers dropped sharply versus your baseline. Keep today light "
            "and prioritise sleep and hydration.",
            "light_recovery",
        )
    if features.fatigue_trend == "declining" or features.risk_score >= 0.55:
        return (
            "Fatigue is trending up. Keep intensity moderate and warm up thoroughly "
            "before any hard efforts.",
            "moderate_drill",
        )
    if features.risk_score <= 0.3:
        action = "moderate_drill" if is_in_season else "heavy_sprint"
        return ("You're well recovered. Good day for a higher-intensity session.", action)
    return ("You're moderately ready. Train as planned but ease off if anything flares up.", "moderate_drill")

File: supabase/functions/_shared/llm.ts
// =============================================================================
// The model chain every Edge Function calls: Groq first, OpenRouter after.
//
// WHY THIS FILE EXISTS. Three Edge Functions each held their own Anthropic
// client, their own model id and their own error handling — while the
// Cloudflare Worker, which serves most of the app's AI, ran an entirely
// different provider chain. Two AI stacks in one product means two sets of
// credentials, two bills, two things to rotate, and features that fail in
// different ways for reasons nobody can hold in their head at once. There is
// now one chain, and it is the Worker's.
//
// GROQ FIRST because it is by a distance the fastest of the two, and every one
// of these calls happens while somebody waits — a photo of a plate, a question
// to the coach, a program being built. OPENROUTER SECOND because breadth is
// exactly what a fallback needs: when a Groq model is retired, rate-limited or
// simply down, the same request goes out to a different company's hardware
// rather than failing.
//
// A rung whose key is missing is SKIPPED, not failed. That is what makes the
// two providers independently optional: set only GROQ_API_KEY and everything
// works with no fallback; set only OPENROUTER_API_KEY and it works a little
// slower. Set neither and the caller gets a 503 that says so.
//
// Both providers speak the OpenAI chat-completions shape, which is the whole
// reason this is one function and not two.
//
// Secrets: GROQ_API_KEY and/or OPENROUTER_API_KEY
// Optional overrides: GROQ_TEXT_MODELS, GROQ_VISION_MODELS,
//                     OPENROUTER_TEXT_MODELS, OPENROUTER_VISION_MODELS
// =============================================================================

export type Provider = "groq" | "openrouter";

export interface Rung {
  provider: Provider;
  model: string;
}

const ENDPOINT: Record<Provider, string> = {
  groq: "https://api.groq.com/openai/v1/chat/completions",
  openrouter: "https://openrouter.ai/api/v1/chat/completions",
};

const KEY_VAR: Record<Provider, string> = {
  groq: "GROQ_API_KEY",
  openrouter: "OPENROUTER_API_KEY",
};

/**
 * Defaults, mirroring what the Worker runs.
 *
 * VISION IS A MUCH SHORTER LIST THAN TEXT, and that asymmetry is the entire
 * reason the meal photo estimator was broken. Most models cannot see. Groq's
 * production chain — gpt-oss-120b, llama-3.3-70b — is text-only, so photos were
 * being posted to something incapable of looking at them and the app blamed the
 * athlete's photography. A model belongs in a vision list only if it has been
 * checked to accept image input.
 *
 * `qwen/qwen3.6-27b` is Groq's multimodal model: 131K context, up to 5 images,
 * 20MB a request, and it takes a base64 data URL in `image_url` exactly like
 * the OpenRouter models below.
 */
const DEFAULTS: Record<Provider, { text: string[]; vision: string[] }> = {
  groq: {
    text: ["openai/gpt-oss-120b", "llama-3.3-70b-versatile"],
    vision: ["qwen/qwen3.6-27b"],
  },
  openrouter: {
    text: ["deepseek/deepseek-chat", "google/gemini-2.5-flash"],
    vision: ["google/gemini-2.5-flash", "openai/gpt-4.1-mini"],
  },
};

function models(provider: Provider, kind: "text" | "vision"): string[] {
  const envVar = `${provider.toUpperCase()}_${kind.toUpperCase()}_MODELS`;
  const configured = (Deno.env.get(envVar) || "").split(",").map((s) => s.trim()).filter(Boolean);
  return configured.length ? configured : DEFAULTS[provider][kind];
}

/**
 * The rungs to try, in order, skipping any provider without a key.
 *
 * Exported so a caller can report an empty chain as "not configured" before
 * doing any work, and so `/health`-style diagnostics can say what is live.
 */
export function chain(kind: "text" | "vision"): Rung[] {
  const out: Rung[] = [];
  for (const provider of ["groq", "openrouter"] as Provider[]) {
    if (!Deno.env.get(KEY_VAR[provider])) continue;
    for (const model of models(provider, kind)) out.push({ provider, model });
  }
  // Same model twice in a row helps nobody.
  return out.filter((r, i, all) => all.findIndex((x) => x.provider === r.provider && x.model === r.model) === i);
}

export interface CompleteOptions {
  system: string;
  user: string;
  /** A `data:image/...;base64,...` URL. Its presence selects the vision chain. */
  image?: string | null;
  maxTokens: number;
  /**
   * Reject a reply that parses to nothing useful and try the next rung.
   *
   * Without this a model that answers with an apology counts as success, and
   * the caller gets "no food found" from a chain that never actually failed —
   * which is indistinguishable, from the outside, from an empty plate.
   */
  validate?: (text: string) => boolean;
}

export interface CompleteResult {
  text: string;
  model: string;
  provider: Provider;
}

/** Thrown when every rung failed. `detail` names each one and why. */
export class ChainError extends Error {
  constructor(readonly detail: string) {
    super(`no model answered — ${detail}`);
  }
}

/**
 * Try each rung until one gives a usable answer.
 *
 * Errors accumulate rather than being swallowed: when the whole chain fails,
 * the caller can say WHICH models failed and how. That matters more than it
 * sounds — the bug this replaced reported a backend problem as a bad photo,
 * and a vaguer message would have been no better.
 */
export async function complete(opts: CompleteOptions): Promise<CompleteResult> {
  const rungs = chain(opts.image ? "vision" : "text");
  if (!rungs.length) throw new ChainError("no provider key configured");

  const content = opts.image
    ? [
        { type: "image_url", image_url: { url: opts.image } },
        { type: "text", text: opts.user },
      ]
    : opts.user;

  const failures: string[] = [];
  for (const rung of rungs) {
    const label = `${rung.provider}/${rung.model}`;
    try {
      const res = await fetch(ENDPOINT[rung.provider], {
        method: "POST",
        headers: {
          Authorization: `Bearer ${Deno.env.get(KEY_VAR[rung.provider])}`,
          "Content-Type": "application/json",
          // OpenRouter attributes usage by these. Harmless to Groq, which
          // ignores headers it doesn't know.
          "HTTP-Referer": "https://pocketathlete.com",
          "X-Title": "PocketAthlete",
        },
        body: JSON.stringify({
          model: rung.model,
          max_tokens: opts.maxTokens,
          messages: [
            { role: "system", content: opts.system },
            { role: "user", content },
          ],
        }),
      });
      if (!res.ok) {
        failures.push(`${label}: ${res.status} ${(await res.text()).slice(0, 160)}`);
        continue;
      }
      const body = await res.json() as { choices?: { message?: { content?: string } }[] };
      const text = body?.choices?.[0]?.message?.content ?? "";
      if (!text.trim()) {
        failures.push(`${label}: empty reply`);
        continue;
      }
      if (opts.validate && !opts.validate(text)) {
        failures.push(`${label}: unusable reply`);
        continue;
      }
      return { text, model: rung.model, provider: rung.provider };
    } catch (e) {
      failures.push(`${label}: ${e instanceof Error ? e.message : String(e)}`);
    }
  }
  throw new ChainError(failures.join("; "));
}
